### PR TITLE
Allow row values in OMA-SQL Queries and improve the iterate blockwise with them

### PR DESCRIPTION
**Pull request by qw3ry**
SIRI-357

# Improve the iterate blockwise using row values

The skip/limit construct is implemented rather slow in the database.
Because we know the value sorted right above the one we want to continue with, we can instead use this value, in conjunction with a total ordering to remove the entities already read from the result set.
This can result in a significant speedup, especially if you have the right indices, and even more so when there are expensive joins involved.

To implement this without too much hassle, we need the ability to use row values:

# Allow the usage of row values in sql constraints

A row value is what you'd call a tuple, and you can use it in SQL constraints like so:
```
... WHERE (column1, column2) > (5, "foo") ...
```
which is equivalent to
```
... WHERE column1 > 5 OR (column1 = 5 AND column2 > "foo") ...
```

A RowValue entity can contain both values and mappings, and you can also mix them.
As a drive-by, you can now also use them to compare columns with each other:
```
oma.select(Foo.class).where(OMA.FILTERS.eq(new RowValue(Foo.COLUMN1), new RowValue(Foo.COLUMN2)))
```
compiling to
```
... WHERE column1 = column2 ...
```
which might help with some more involved queries

### CODE REVIEW

**Review by andyHa — APPROVED** [review]

File: src/main/java/sirius/db/jdbc/constraints/RowValue.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+ * An n-tuple of normal values that can be used in the OMA {@link sirius.db.jdbc.OMA#FILTERS filters}.
+ * <p>
+ * Use {@link Mapping} objects to indicate a column; every other object will be treated as an actual value.
+ * Usage example:<br>
+ * {@code
+ * oma.select(Entity.class).where(OMA.FILTERS.gte(new RowValue(Entity.COLUMN1, Entity.COLUMN2), new RowValue(4, "Test")))
+ * }
+ * <br>will compile to<br>
+ * {@code
+ * SELECT * FROM Entity e1 WHERE (e1.column1, e2.column2) >= (4, "Test")
```
Comment (qw3ry) [inline src/main/java/sirius/db/jdbc/constraints/RowValue.java:33]: Gibts was vergleichbares für Mongo/Elastic? Weil dann könnte man das ja gleich für alle machen.

**Review by mkeckmkeck — APPROVED** [review]

File: src/main/java/sirius/db/jdbc/constraints/FieldOperator.java [hunk 2]
```diff
@@ -9,32 +9,37 @@
...
 /**
  * Represents a simple field operator as constraint.
  */
 class FieldOperator extends SQLConstraint {
 
-    private Mapping field;
-    private Object value;
-    private String op;
+    private final RowValue lhs;
+    private final RowValue rhs;
```
Comment (sabieber) [inline src/main/java/sirius/db/jdbc/constraints/FieldOperator.java:23]: Optional: Können wir da evtl deskriptivere Feldnamen etc wählen. Also ich vermute mal das steht für left und right hand side oder? Aber so richtig sicher ist man sich da dann nach ner Weile evtl net mehr wenn man da in nem halben Jahr was dran basteln muss.

File: src/main/java/sirius/db/jdbc/constraints/SQLFilterFactory.java [hunk 3]
```diff
@@ -32,22 +32,96 @@ protected Object customTransform(Object value) {
...
+    }
+
+    /**
+     * Represents {@code lhs = rhs} as constraint
+     *
+     * @param lhs the first row value
+     * @param rhs the second row value
+     * @return the generated constraint
+     */
+    public SQLConstraint eq(RowValue lhs, RowValue rhs) {
```
Comment (sabieber) [inline src/main/java/sirius/db/jdbc/constraints/SQLFilterFactory.java:45]: Optional: hier auch :)

File: src/main/java/sirius/db/jdbc/SmartQuery.java [hunk 4]
```diff
@@ -250,102 +256,71 @@ public void iterateBlockwise(Predicate<E> handler) {
...
+        // no last value => we query everything
+        if (lastValue == null) {
+            return query;
+        }
 
-                return true;
-            });
+        // create and install the filter
+        Object[] lhs = new Object[query.orderBys.size()];
+        Object[] rhs = new Object[query.orderBys.size()];
```
Comment (sabieber) [inline src/main/java/sirius/db/jdbc/SmartQuery.java:309]: Optional: hier auch

**Review by sabieber — APPROVED** [review]

**Review by ymo-sci — APPROVED** [review]

**Pull request merged**